---
title: "Sex differences on lung cancer"
output: html_document
date: "2023-01-25"
---

We start by loading some necessary stuff to run the notebook. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
setwd("/home/soel/Desktop/cobra/")
library("cluster")
library("clusterProfiler")
library("data.table")
library('DESeq2')
library("fastDummies")
library("flashClust")
library("recount3")
library("WGCNA")
library(GOstats)
library(org.Hs.eg.db)
library(AnnotationDbi)
library(KEGGandMetacoreDzPathwaysGEO)
library(fgsea)
library(ggplot2)
source('./algorithm.R')

# Extracting data from TCGA
data <- recount3::create_rse_manual(
  project = "THCA",
  project_home = "data_sources/tcga",
  organism = "human",
  annotation = "gencode_v26",
  type = "gene"
)
  
# Gene expression pre-processing, for more information see 
# https://github.com/netZoo/netbooks/blob/main/netbooks/netZooR/gene_expression_for_coexpression_nets.ipynb
G <- transform_counts(data, by = "mapped_reads")
G <- G[data@rowRanges@elementMetadata@listData$gene_type == "protein_coding",]
G <- G[-which(rowSums(G) <= 1),] # Filtering: remove genes with no counts
countMat=SummarizedExperiment::assay(DESeqDataSetFromMatrix(G, data.frame(row.names=seq_len(ncol(G))), ~1), 1)
vsd <- vst(countMat, blind=FALSE)
  
# Extract metadata that we want to include in COBRA
metadata_url <- locate_url(
  "THCA",
  "data_sources/tcga")
metadata <- read_metadata(file_retrieve(url = metadata_url))
  
sex <- metadata$tcga.gdc_cases.demographic.gender
race <- metadata$tcga.gdc_cases.demographic.race
stage <- metadata$tcga.gdc_cases.diagnoses.tumor_stage
batch <- metadata$tcga.cgc_case_batch_number
age <- metadata$tcga.cgc_case_age_at_diagnosis
nat <- metadata$tcga.gdc_cases.samples.sample_type
nat <- ifelse(nat == "Solid Tissue Normal", 0, 1)

X <- cbind(rep(1, dim(G)[2]), nat, sex == 'female', age, as.matrix(dummy_cols(race)[-1]), as.matrix(dummy_cols(stage)[-1]), as.matrix(dummy_cols(batch)[-1]))
cobra <- themethod(X, vsd, absolute = F, eigen_function = eigs_sym)
Sigma_D <- cobra$Q %*% diag(cobra$estimates[2,]) %*% t(cobra$Q)
rownames(Sigma_D) <- rownames(vsd)
colnames(Sigma_D) <- rownames(vsd)
```

```{python}
import netZooPy.panda.calculations as nz
import numpy as np
import pandas as pd
Sigma_D = r.Sigma_D
cache=True
ppi = pd.read_csv("tissues_ppi.txt", sep = '\t', header = None)
motif = pd.read_csv("tissues_motif.txt", sep = '\t', header = None)
motif_filter = motif[motif[2] != 0]

gene_map = {}
for g in Sigma_D.columns:
  gene_map[g] = g[0:15]
Sigma_D = Sigma_D.rename(columns=gene_map)
c = 0
count_map = {}
for g in Sigma_D.columns:
  count_map[c] = g
  c += 1
Sigma_D = Sigma_D.rename(index=count_map)

if not cache:
  ppi_matrix = ppi_matrix = pd.DataFrame(0, index = list(set(ppi[0])), columns = list(set(ppi[0])))
  for i in range(ppi.shape[0]): 
    ppi_matrix[ppi.iloc[i, 0]][ppi.iloc[i, 1]] = ppi.iloc[i, 2]
    ppi_matrix[ppi.iloc[i, 1]][ppi.iloc[i, 0]] = ppi.iloc[i, 2]
  ppi_matrix.to_csv("ppi_matrix.csv")
  
  motif_matrix = pd.DataFrame(0, index = Sigma_D.index, columns = ppi_matrix.index)
  for i in range(motif_filter.shape[0]):
    if i % 10000 == 0:
      print(i)
    tf = motif_filter.iloc[i, 0]
    g = motif_filter.iloc[i, 1]
    if g in motif_matrix.index and tf in motif_matrix.columns:
      motif_matrix[tf][g] = 1
  motif_matrix.to_csv("motif_matrix.csv")
else:
  ppi_matrix=pd.read_csv("ppi_matrix.csv", index_col=0)
  motif_matrix=pd.read_csv("motif_matrix.csv", index_col=0)

tf_to_remove = []
for p in motif_matrix.columns:
  if np.sum(motif_matrix[p].to_numpy()) == 0:
    tf_to_remove.append(p)
genes_to_remove = []
for i in range(motif_matrix.shape[0]):
  if np.sum(motif_matrix.iloc[i,].to_numpy()) == 0:
    genes_to_remove.append(motif_matrix.index[i])
Sigma_D.drop(genes_to_remove, inplace=True)
Sigma_D.drop(genes_to_remove, axis = 1, inplace=True)
motif_matrix.drop(genes_to_remove, inplace=True)
motif_matrix.drop(tf_to_remove, axis=1, inplace=True)
ppi_matrix.drop(tf_to_remove,inplace=True)
ppi_matrix.drop(tf_to_remove, axis=1, inplace=True)

R = nz.compute_panda_cpu(Sigma_D.to_numpy(), ppi_matrix.to_numpy(), np.float_(motif_matrix.T.to_numpy()))
pd.DataFrame(R, index=ppi_matrix.index, columns=Sigma_D.index).to_csv("R_D.csv")
```

```{r}
analysis <- function(diff){
  pathways <- gmtPathways("/experiments/tcga/nat/cobra_diff/kegg.gmt")
  fgseaRes <- fgsea(pathways, diff, minSize=15, maxSize=500, nperm=10000)
  sig <- fgseaRes[fgseaRes$padj < 0.1,]
  sig$pathway[sig$NES > 0][1:10]
  sig$pathway[sig$NES < 0][1:10]
  dat <- data.frame(fgseaRes)
  # Settings
  fdrcut <- 0.05 # FDR cut-off to use as output for significant signatures
  dencol_neg <- "blue" # bubble plot color for negative ES
  dencol_pos <- "red" # bubble plot color for positive ES
  signnamelength <- 4 # set to remove prefix from signature names (2 for "GO", 4 for "KEGG", 8 for "REACTOME")
  asp <- 3 # aspect ratio of bubble plot
  charcut <- 100 # cut signature name in heatmap to this nr of characters
  # Make signature names more readable
  a <- as.character(dat$pathway) # 'a' is a great variable name to substitute row names with something more readable
  for (j in 1:length(a)){
    a[j] <- substr(a[j], signnamelength+2, nchar(a[j]))
  }
  a <- tolower(a) # convert to lower case (you may want to comment this out, it really depends on what signatures you are looking at, c6 signatures contain   gene names, and converting those to lower case may be confusing)
  for (j in 1:length(a)){
    if(nchar(a[j])>charcut) { a[j] <- paste(substr(a[j], 1, charcut), "...", sep=" ")}
  } # cut signature names that have more characters than charcut, and add "..."
  a <- gsub("_", " ", a)
  dat$NAME <- a
  # Determine what signatures to plot (based on FDR cut)
  dat2 <- dat[dat[,"padj"]<fdrcut,]
  dat2 <- dat2[order(dat2[,"padj"]),] 
  dat2$signature <- factor(dat2$NAME, rev(as.character(dat2$NAME)))
  # Determine what labels to color
  sign_neg <- which(dat2[,"NES"]<0)
  sign_pos <- which(dat2[,"NES"]>0)
  # Color labels
  signcol <- rep(NA, length(dat2$signature))
  signcol[sign_neg] <- dencol_neg # text color of negative signatures
  signcol[sign_pos] <- dencol_pos # text color of positive signatures
  signcol <- rev(signcol) # need to revert vector of colors, because ggplot starts plotting these from below
  # Plot bubble plot
  png("/experiments/tcga/nat/cobra_diff/cobra_diff.jpg")
  g<-ggplot(dat2, aes(x=padj,y=signature,size=size))
  g+geom_point(aes(fill=NES), shape=21, colour="white")+
    theme_bw()+ # white background, needs to be placed before the "signcol" line
    xlim(0,fdrcut)+
    scale_size_area(max_size=10,guide="none")+
    scale_fill_gradient2(low=dencol_neg, high=dencol_pos)+
    theme(axis.text.y = element_text(colour=signcol))+
    theme(aspect.ratio=asp, axis.title.y=element_blank()) # test aspect.ratio
  dev.off()
}
```

```{r}
R_D <- fread("experiments/tcga/nat/cobra_diff/R_D.csv")
rownames(R_D) <- R_D$V1
R_D$V1 <- NULL

annot <- read.csv("experiments/tcga/nat/cobra_diff/gen_v26_mapping.csv")
a <- c()
for(g in annot$gene_id){
  a <- cbind(a, substr(g, 0, 15))
}
annot$gene_id <- t(a)
geneId <- annot[match(colnames(R_D),annot$gene_id),"gene_name"]
colnames(R_D) <- geneId
analysys(colSums(R_D))
```
